src/cross_mapping_mpi_bench.cc

```diff
@@ -0,0 +1,285 @@
+#include <iostream>
+
+#include <argh.h>
+
+#include "cross_mapping_cpu.h"
+#include "dataframe.h"
+#include "embedding_dim_cpu.h"
+#include "mpi_master.h"
+#include "mpi_worker.h"
+#ifdef ENABLE_GPU_KERNEL
+#include "cross_mapping_gpu.h"
+#include "embedding_dim_gpu.h"
+#endif
+#include "stats.h"
+#include "timer.h"
+
+class EmbeddingDimMPIMaster : public MPIMaster
+{
+public:
+    std::vector<uint32_t> optimal_E;
+
+    EmbeddingDimMPIMaster(const DataFrame df, MPI_Comm comm)
+        : MPIMaster(comm), current_id(0), dataframe(df)
+    {
+        optimal_E.resize(df.n_columns());
+    }
+    ~EmbeddingDimMPIMaster() {}
+
+protected:
+    uint32_t current_id;
+    DataFrame dataframe;
+
+    void next_task(nlohmann::json &task) override
+    {
+        task["id"] = current_id;
+        current_id++;
+    }
+
+    bool task_left() const override
+    {
+        return current_id < dataframe.n_columns();
+    }
+
+    void task_done(const nlohmann::json &result) override
+    {
+        std::cout << "Timeseries #" << result["id"] << " best E=" << result["E"]
+                  << std::endl;
+
+        optimal_E[result["id"]] = result["E"];
+    }
+};
+
+template <class T> class EmbeddingDimMPIWorker : public MPIWorker
+{
+public:
+    EmbeddingDimMPIWorker(const DataFrame df, bool verbose, MPI_Comm comm)
+        : MPIWorker(comm),
+          embedding_dim(std::unique_ptr<EmbeddingDim>(new T(20, 1, 1, true))),
+          dataframe(df), verbose(verbose)
+    {
+    }
+    ~EmbeddingDimMPIWorker() {}
+
+protected:
+    std::unique_ptr<EmbeddingDim> embedding_dim;
+    DataFrame dataframe;
+    bool verbose;
+
+    void do_task(nlohmann::json &result, const nlohmann::json &task) override
+    {
+        const auto id = task["id"];
+        const auto ts = dataframe.columns[id];
+        const auto best_E = embedding_dim->run(ts);
+
+        if (verbose) {
+            std::cout << "Find embedding dimension for column #" << id
```

Comment:
You don't need to log from the workers since you are logging in the master (task_done). Besides, the outputs from all the processes will be scrambled together.